pgraph/pgraph.go

```diff
@@ -246,17 +255,67 @@ func (vs VertexSlice) Less(i, j int) bool { return vs[i].String() < vs[j].String
 // Sort is a convenience method.
 func (vs VertexSlice) Sort() { sort.Sort(vs) }
 
+// VertexIntenseSlice is a linear list of vertices. It can be sorted.
+type VertexIntenseSlice struct {
+	g        *Graph
+	vertices []Vertex
+}
+
+func (vs VertexIntenseSlice) Len() int {
+	return len(vs.vertices)
+}
+
+func (vs VertexIntenseSlice) Swap(i, j int) {
+	v := vs.vertices
+	v[i], v[j] = v[j], v[i]
+}
+
+func (vs VertexIntenseSlice) Less(i, j int) bool {
+	v := vs.vertices
+	if v[i].String() < v[j].String() {
+		return true
+	} else if v[i].String() > v[j].String() {
+		return false
+	}
+
+	// Labels are equal, time to check neighbor labels
+	neighSet1 := vs.g.GraphVertices(v[i])
+	neighSet2 := vs.g.GraphVertices(v[j])
+
+	if len(neighSet1) < len(neighSet2) {
+		return true
+	}
+
+	sort.Sort(VertexSlice(neighSet1))
+	sort.Sort(VertexSlice(neighSet2))
+
+	// TODO: maybe building a hash here and compare the result is better?
+	for i := 0; i < len(neighSet1); i++ {
+		if neighSet1[i].String() < neighSet2[i].String() {
+			return true
+		}
+	}
+
+	return false
+}
+
 // VerticesSorted returns a sorted slice of all vertices in the graph.
 // The order is sorted by String() to avoid the non-determinism in the map type.
 func (g *Graph) VerticesSorted() []Vertex {
-	var vertices []Vertex
```

Comment:
Not sure if this change is _really_ necessary, but if it is, let's do it in a separate commit, since it has nothing to do with this commit which is not trivial, so we don't want to combine them if there's ever something we learn we need to revert or change in this commit.